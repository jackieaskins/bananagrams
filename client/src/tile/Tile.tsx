import { TILE_SIZE } from './constants';

const Tile = ({ letter }: { letter: string }): JSX.Element => (
  <div
    className="no-drag"
    css={{
      width: `${TILE_SIZE}px`,
      height: `${TILE_SIZE}px`,
      backgroundColor: '#ffffc7',
      border: '0.5px solid black',
      borderRadius: '5px',
      flexShrink: 0,
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
    }}
  >
    {letter}
  </div>
);

export default Tile;
