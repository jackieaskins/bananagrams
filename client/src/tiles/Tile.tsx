/* eslint-disable @typescript-eslint/ban-ts-comment */
import { Typography, Box } from '@material-ui/core';
import { useDrag } from 'react-dnd';

import { Tile as TileType } from './types';

type TileProps = {
  boardLocation: { x: number; y: number } | null;
  tile: TileType;
  color?: string;
};

const Tile = ({
  boardLocation,
  tile: { id, letter },
  color = 'black',
}: TileProps): JSX.Element => {
  const [{ isDragging }, dragRef] = useDrag({
    item: { type: 'TILE', id, boardLocation },
    collect: (monitor) => ({
      isDragging: monitor.isDragging(),
    }),
  });

  const margin = !!boardLocation ? '0' : '5px';

  return (
    <Box
      // @ts-ignore
      ref={dragRef}
      display="inline-flex"
      justifyContent="center"
      alignItems="center"
      border={1}
      borderRadius="borderRadius"
      style={{
        color: color,
        backgroundColor: '#ffffc7',
        cursor: 'move',
        height: '23px',
        margin,
        opacity: isDragging ? 0.5 : 1,
        width: '23px',
      }}
    >
      <Typography variant="body2">{letter}</Typography>
    </Box>
  );
};

export default Tile;
