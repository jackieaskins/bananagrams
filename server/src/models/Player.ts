import Hand, { HandJSON } from './Hand';
import Board, { BoardLocation, BoardJSON } from './Board';
import BaseModel from './BaseModel';

export type PlayerJSON = {
  userId: string;
  username: string;
  isReady: boolean;
  isTopBanana: boolean;
  gamesWon: number;
  hand: HandJSON;
  board: BoardJSON;
};

export default class Player implements BaseModel<PlayerJSON> {
  private userId: string;
  private username: string;
  private ready = false;
  private topBanana = false;
  private gamesWon = 0;
  private hand = new Hand();
  private board = new Board();

  constructor(userId: string, username: string) {
    this.userId = userId;
    this.username = username;
  }

  getUserId(): string {
    return this.userId;
  }

  getUsername(): string {
    return this.username;
  }

  isReady(): boolean {
    return this.ready;
  }

  setReady(ready: boolean): void {
    this.ready = ready;
  }

  isTopBanana(): boolean {
    return this.topBanana;
  }

  setTopBanana(topBanana: boolean): void {
    this.topBanana = topBanana;
  }

  getGamesWon(): number {
    return this.gamesWon;
  }

  incrementGamesWon(): void {
    this.gamesWon++;
  }

  getHand(): Hand {
    return this.hand;
  }

  getBoard(): Board {
    return this.board;
  }

  toJSON(): PlayerJSON {
    const { userId, username, ready, topBanana, gamesWon, board, hand } = this;

    return {
      userId,
      username,
      gamesWon,
      isReady: ready,
      isTopBanana: topBanana,
      board: board.toJSON(),
      hand: hand.toJSON(),
    };
  }

  reset(): void {
    this.hand.reset();
    this.board.reset();
  }

  moveTileFromHandToBoard(tileId: string, location: BoardLocation): void {
    this.board.validateEmptySquare(location);
    const tile = this.hand.removeTile(tileId);
    this.board.addTile(location, tile);
  }

  moveTileFromBoardToHand(location: BoardLocation): void {
    const tile = this.board.removeTile(location);
    this.hand.addTiles([tile]);
  }

  moveTileOnBoard(from: BoardLocation, to: BoardLocation): void {
    this.board.validateEmptySquare(to);
    const tile = this.board.removeTile(from);
    this.board.addTile(to, tile);
  }
}
