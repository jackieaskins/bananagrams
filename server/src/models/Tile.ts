import BaseModel from './BaseModel';

export type TileJSON = {
  id: string;
  letter: string;
};

export default class Tile implements BaseModel<TileJSON> {
  private id: string;
  private letter: string;

  constructor(id: string, letter: string) {
    this.id = id;
    this.letter = letter;
  }

  getId(): string {
    return this.id;
  }

  getLetter(): string {
    return this.letter;
  }

  toJSON(): TileJSON {
    const { id, letter } = this;
    return { id, letter };
  }

  reset(): void {
    return;
  }
}
