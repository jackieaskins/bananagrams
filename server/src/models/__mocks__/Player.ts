export default jest.fn().mockImplementation((userId, username) => ({
  getUserId: jest.fn().mockReturnValue(userId),
  getUsername: jest.fn().mockReturnValue(username),
  toJSON: jest.fn().mockReturnValue('PlayerJSON'),
  reset: jest.fn(),
}));
