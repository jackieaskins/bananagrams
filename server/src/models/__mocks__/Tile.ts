export default jest.fn().mockImplementation((id, letter) => ({
  getId: jest.fn().mockReturnValue(id),
  getLetter: jest.fn().mockReturnValue(letter),
  toJSON: jest.fn().mockReturnValue(`TileJSON - ${id} ${letter}`),
}));
