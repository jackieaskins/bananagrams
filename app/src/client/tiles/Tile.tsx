import { Box, Text } from "@chakra-ui/react";
import { useDrag } from "react-dnd";
import { Tile as TileType } from "@/types/tile";

type TileProps = {
  boardLocation: { x: number; y: number } | null;
  tile: TileType;
  color?: string;
};

export default function Tile({
  boardLocation,
  tile: { id, letter },
  color = "black",
}: TileProps): JSX.Element {
  const [{ isDragging }, dragRef] = useDrag({
    item: { type: "TILE", id, boardLocation },
    collect: (monitor) => ({
      isDragging: monitor.isDragging(),
    }),
  });

  const margin = boardLocation ? "0" : "5px";

  return (
    <Box
      ref={dragRef}
      alignItems="center"
      backgroundColor="yellow.100"
      border="1px"
      borderColor={color}
      borderRadius="5px"
      color={color}
      cursor="move"
      display="inline-flex"
      height="25px"
      justifyContent="center"
      margin={margin}
      opacity={isDragging ? 0.5 : 1}
      width="25px"
    >
      <Text as="b">{letter}</Text>
    </Box>
  );
}
