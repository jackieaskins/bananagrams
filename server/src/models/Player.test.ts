import Player from './Player';
import Tile from './Tile';

jest.mock('./Tile');
jest.mock('./Board');
jest.mock('./Hand');

describe('Player Model', () => {
  const userId = 'userId';
  const username = 'username';

  let player: Player;

  beforeEach(() => {
    player = new Player(userId, username);
  });

  describe('getUserId', () => {
    test('returns user id', () => {
      expect(player.getUserId()).toEqual(userId);
    });
  });

  describe('getUsername', () => {
    test('returns user name', () => {
      expect(player.getUsername()).toEqual(username);
    });
  });

  describe('set/isReady', () => {
    test('returns false by default', () => {
      expect(player.isReady()).toEqual(false);
    });

    test('sets ready', () => {
      player.setReady(true);

      expect(player.isReady()).toEqual(true);
    });
  });

  describe('set/isTopBanana', () => {
    test('returns false by default', () => {
      expect(player.isTopBanana()).toEqual(false);
    });

    test('sets isTopBanana', () => {
      player.setTopBanana(true);

      expect(player.isTopBanana()).toEqual(true);
    });
  });

  describe('get/incrementGamesWon', () => {
    test('returns 0 by default', () => {
      expect(player.getGamesWon()).toEqual(0);
    });

    test('increments games won', () => {
      player.incrementGamesWon();

      expect(player.getGamesWon()).toEqual(1);
    });
  });

  describe('getHand', () => {
    test('returns an initialized hand', () => {
      expect(player.getHand()).toMatchSnapshot();
    });
  });

  describe('getBoard', () => {
    test('returns an initialized board', () => {
      expect(player.getBoard()).toMatchSnapshot();
    });
  });

  describe('toJSON', () => {
    test('converts fields into JSON', () => {
      expect(player.toJSON()).toMatchSnapshot();
    });
  });

  describe('reset', () => {
    beforeEach(() => {
      player.reset();
    });

    test('resets hand', () => {
      expect(player.getHand().reset).toHaveBeenCalledWith();
    });

    test('resets board', () => {
      expect(player.getBoard().reset).toHaveBeenCalledWith();
    });
  });

  describe('moveTileFromHandToBoard', () => {
    const tile = new Tile('A1', 'A');
    const location = { x: 0, y: 0 };

    beforeEach(() => {
      player.getHand().addTiles([tile]);
      player.moveTileFromHandToBoard(tile.getId(), location);
    });

    test('validates that board square is empty', () => {
      expect(player.getBoard().validateEmptySquare).toHaveBeenCalledWith(
        location
      );
    });

    test('removes tile from hand', () => {
      expect(player.getHand().removeTile).toHaveBeenCalledWith(tile.getId());
    });

    test('adds removed tile to board', () => {
      expect(player.getBoard().addTile).toHaveBeenCalledWith(location, tile);
    });
  });

  describe('moveTileFromBoardToHand', () => {
    const location = { x: 0, y: 0 };
    const tile = new Tile('A1', 'A');

    beforeEach(() => {
      player.getBoard().addTile(location, tile);
      player.moveTileFromBoardToHand(location);
    });

    test('removes tile from board', () => {
      expect(player.getBoard().removeTile).toHaveBeenCalledWith(location);
    });

    test('adds removed tile to hand', () => {
      expect(player.getHand().addTiles).toHaveBeenCalledWith([tile]);
    });
  });

  describe('moveTileOnBoard', () => {
    const from = { x: 0, y: 0 };
    const to = { x: 1, y: 1 };
    const tile = new Tile('A1', 'A');

    beforeEach(() => {
      player.getBoard().addTile(from, tile);
      player.moveTileOnBoard(from, to);
    });

    test('validates to location is empty', () => {
      expect(player.getBoard().validateEmptySquare).toHaveBeenCalledWith(to);
    });

    test('removes tile from from location', () => {
      expect(player.getBoard().removeTile).toHaveBeenCalledWith(from);
    });

    test('adds removed tile to to location', () => {
      expect(player.getBoard().addTile).toHaveBeenCalledWith(to, tile);
    });
  });
});
