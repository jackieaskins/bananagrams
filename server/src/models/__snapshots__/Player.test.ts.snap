// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player Model getBoard returns an initialized board 1`] = `"BoardJSON"`;

exports[`Player Model getHand returns an initialized hand 1`] = `"HandJSON"`;

exports[`Player Model toJSON converts fields into JSON 1`] = `
Object {
  "board": "BoardJSON",
  "gamesWon": 0,
  "hand": "HandJSON",
  "isReady": false,
  "isTopBanana": false,
  "userId": "userId",
  "username": "username",
}
`;
